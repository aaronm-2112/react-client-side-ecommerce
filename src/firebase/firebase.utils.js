import firebase from "firebase/app";
import "firebase/firestore";
import "firebase/auth";

const config = {
  apiKey: "AIzaSyAPurqbdFlpp2mtr8cKQX7ZAAfoy9HV7Yw",
  authDomain: "ecommerce-crwn-33e21.firebaseapp.com",
  projectId: "ecommerce-crwn-33e21",
  storageBucket: "ecommerce-crwn-33e21.appspot.com",
  messagingSenderId: "1078485096954",
  appId: "1:1078485096954:web:b9bf49c04eae40b5e5ac43",
  measurementId: "G-EML7624QZF",
};

firebase.initializeApp(config);

export const auth = firebase.auth();
export const firestore = firebase.firestore();

const provider = new firebase.auth.GoogleAuthProvider();
provider.setCustomParameters({ prompt: "select_account" });
export const signInWithGoogle = () => auth.signInWithPopup(provider);

export default firebase;
