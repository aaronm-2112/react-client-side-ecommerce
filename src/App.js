import React from "react";
import HomePage from "./pages/homepage/homepage.component";

import { Switch, Route, Redirect } from "react-router-dom";
import "./App.css";
import ShopPage from "./pages/shop/shop.component";
import Header from "./components/header/header.component";
import SignInAndSignUpPage from "./pages/signin-and-signup/signin-and-signup.component";
import { auth, createUserProfileDocument } from "./firebase/firebase.utils";
import { connect } from "react-redux";
import { setCurrentUser } from "./redux/user/user.actions";
import { selectCurrentUser } from "./redux/user/user.selector";
import { createStructuredSelector } from "reselect";
import CheckoutPage from "./pages/checkout/checkout.component";

class App extends React.Component {
  // set up a subscription to Google Firebase for Auth status changes -- this persists while the client is on our site
  componentDidMount() {
    const { setCurrentUser } = this.props;
    this.unsubscribeFromAuth = auth.onAuthStateChanged(async (userAuth) => {
      // ignore if the user signed out -- we are alerted both when a user signs in and out. On sign out there is a null userAuth object returned
      if (userAuth) {
        // the user profile document is the documentReference to the collection table for the current user
        const userRef = await createUserProfileDocument(userAuth);

        // whenever we change the user's data using the userSnapshot update the state of our app to match the user's firestore data
        userRef.onSnapshot((snapShot) => {
          setCurrentUser({
            id: snapShot.id,
            ...snapShot.data(),
          });
        });
      } else {
        // if user is signed out set currentUser to null
        setCurrentUser(userAuth);
      }
    });
  }

  // stores the subscription for Auth status changes
  unsubscribeFromAuth = null;

  // when the app closes unsubscribe fron the Auth subscription to avoid memory leaks
  componentWillUnmount() {
    this.unsubscribeFromAuth();
  }

  render() {
    const { currentUser } = this.props;
    return (
      <div>
        <Header />
        <Switch>
          <Route exact path="/" component={HomePage} />
          <Route exact path="/checkout" component={CheckoutPage} />
          <Route path="/shop" component={ShopPage} />

          <Route
            exact
            path="/signin"
            render={() =>
              currentUser ? <Redirect to="/" /> : <SignInAndSignUpPage />
            }
          />
        </Switch>
      </div>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
});

const mapDispatchToProps = (dispatch) => {
  return { setCurrentUser: (user) => dispatch(setCurrentUser(user)) };
};

// connect returns a function. Then we execute this function with App as a parameter
export default connect(mapStateToProps, mapDispatchToProps)(App);
