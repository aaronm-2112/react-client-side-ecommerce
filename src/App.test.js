import { shallow } from "enzyme";
import App from "./App";
import * as fireUtils from "./firebase/firebase.utils";

// make a mock of the onAuthStateChanged method
fireUtils.auth.onAuthStateChanged = jest.fn();

describe("Tests for the App component", () => {
  let wrapper;
  beforeEach(() => {
    // create a wrapper
    wrapper = shallow(<App />);
  });

  it("Should have a matching wrapper", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Should launch componentDidMount", () => {
    const instance = wrapper.instance();

    instance.componentDidMount();

    expect(fireUtils.auth.onAuthStateChanged).toHaveBeenCalled();
  });
});
