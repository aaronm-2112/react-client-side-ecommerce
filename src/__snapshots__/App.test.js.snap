// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the App component Should have a matching wrapper 1`] = `<MainPage />`;
